### DISCUSSION TITLE
Low constant current to pre-tension drivetrain?

### DISCUSSION THREAD

**Discussion by EigenMania** [danielnilsson9/bbs-fw#74] [Q&A]
Sometimes when applying throttle / PAS during starts from rest I experience a sharp clunk before the drivetrain engages. I assume that this is from a combination of play in the motor gearbox and pawl spacing in my rear wheel. I'm wondering if a very low current to the motor would be able to "pre-tension" the system (without driving the bike), so that the motor and freewheel are constantly engaged.  The throttle ADC is 8-bit, so with a max current of 30A, I could adjust the throttle lower limit to put a constant ~100-200mA through the motor. Has anyone ever tried this? Maybe someone can already tell me if it's a really bad idea :P

**Comment by dav0000000** [comment]
I think it's a great idea and maybe just a brake and lower speed limit cutoff like at 8km/h or something. I would probably not adjust it from the throttle and use apply_ramp code to maintain the current until brakes or under speed is detected.

**Reply by dav0000000** [reply]
And yes of course anything that keeps the motor running can be dangerous, especially without brake sensors.

**Reply by EigenMania** [reply]
Hey @dav0000000! Thanks a bunch for your suggestions :) I finally got around to playing with the code this weekend, and I have a nice solution implemented now. I have a pretension function that sets a 1% motor current setpoint if the road-speed is above a certain threshold (configurable). I put this block before everything (throttle, pas, cruise, ramps, thermal limits, etc...), so that it doesn't interfere with anything else. The road speed cutoff was certainly a good idea! Turns out that 1% of 30A max is still enough to drive my bike on it's own from rest :P It's also low enough not to interfere with shifting my gear hub, so once you are moving it works great and it's barely noticeable. And on a freewheel system it might even be cooler since you'd be able to shift and you wouldn't need to pedal to finish the gear change.

**Reply by dav0000000** [reply]
I played with it too. It works great and I'll stick with it. It has its dangers but they are outweighed by the smooth take off after stopping pedalling. I also went with 1% and it certainly keeps the bike going, it's definitely hitting a controller minimum current which is higher than 1% of 30A. I only use pas and have brake sensors.

**Comment by stts2** [comment]
Im going to make a torque arm for my motor that has a stiff spring on it. That way when things go clunk, the spring is a shock absorber. I got no clue why these motors are being locked down dead tight. All it does is make things break in time.

It looks like a ton of great features are already in this new code. But one thing I will eventually do is put my name in the code. And have the control post my name on the display when its powered up. And thieves wont be able to grind it off. Replace the display and my name still pops up. They wont have a clue whats going on. Also maybe even a message. Tracker says the cops are near. Hah. Well maybe not that message. I wouldn't want them to go crazy and throw my bike in the river. Open source is so sweet.

Is anybody doing open source on displays? They got really nice touch screen displays now. For tons of data entry to run some fancy features.

**Comment by kklmm** [comment]
Hi @EigenMania @dav0000000 

I want to add some feedback and feature request for pre-tension implementation. 
Am testing implementation from @EigenMania firmware for about two weeks now of daily riding. 

- in current state pre-tension is implemented by spinning the chainring at a constant consistent RPM.  
Pre-tension works well on a flat terrain with calm riding but completely fails when using throttle aggressively on an uneven mountain like terrain. Riding in a quick succession of acceleration then breaking then acceleration again constant consistent spinning of the chainring has no time to catch up with the drivetrain. 

- when using throttle aggressively rpm of the freewheel does not match rpm of the chainring due to the uneven momentum of the rotation. a momentary empty unengaged gap in drive train is formed and activating throttle before slow rpm of pre-tension can pick up the slack causes drive train to aggressively slam into the freewheel pinions. pre-tension by a fixed static consistent constant RPM then does not work at all. 

BBSHD, 52V 33A, 36T front chainring.

I propose an improvement to pretension feature:
- supplement existing fixed static constant consistent RPM of the chainring at low current (fixed rpm) with a dynamic pre-tension implementation on throttle activation. achieve this by delaying power output to the engine to minimum power just long enough to tension up the drive train. 

- to identify the moment when the drive train engages spin up the engine at decent rpm (higher than current static rpm) with minimum power. a) measure engine rpm. when drive train engages engine rpm will suddenly drop due to engine not having enough power to sustain now engaged drive train at minimum current. Or b) measure engine power consumption. spin the engine up to a certain rpm and modest power. when drive train engages power consumption will increase to reach the target rpm. At this point send power to the engine normally.

- This implementation of pre-tension should theoretically require no more than a half a revolution of the chainring and be engaged in under a second.  2 seconds absolute max. 

- Make this pre-tension an active functionality of the throttle whenever power consumption is below a certain limit (below 10%?) so the drive train always engages based on power consumption regardless of speed or rpm.  an unengaged engine is likely to have near zero power consumption (even at decent rpm due to momentum!) while engaged engine has considerable power consumption.

- this does not account for PAS. when pedalling this pre-tension is unnecessary as it is assumed the pedal rotation engages the drivetrain and implementation of PAS+throttle pretension complicates things excessively.

Please let me know your thoughts.

**Reply by dav0000000** [reply]
We cannot measure motor rpm. I have tried to find a magic opcode but it's been a waste of time.

**Reply by danielnilsson9** [reply]
@dav0000000 If I remember correctly from tracing the PCB the hall sensors are also connected to the STC controller, so it should be possible to measure motor rpm. I just didn't implement it since it wasn't needed at the time.

**Reply by danielnilsson9** [reply]
According to my drawing they are on STC P3.4, P5.0, P0.6

**Reply by dav0000000** [reply]
It certainly would be interesting to have the motor speed but if the time to read this information was long then I guess it would have to fit in with the ramp up/down and status/current/voltage reading operations. I still feel we could use the motor cadence percent limit better (OPCODE_TARGET_SPEED), including setting it initially low in the ramp up to help with this drivetrain pre tensioning, as it really is very smooth at limiting the motor. I've been meaning to try something along those lines but I envisage drivetrain recoil problems during ramp up when in the big rear cogs as the pre tension is completely different. In low gears the engine must spin up more to catch the drivetrain. It could be used to limit speed also if only we knew the given motor speed to set it correctly... if that makes sense. 

I've settled for increasing the initial ramp up current based on speed with a slightly longer delay and it works ok. My motor is always spinning way faster than I am pedalling which is the same use case as the throttle only users. Do you think that knowing the motor speed would be useful or more a can of worms?

**Reply by kklmm** [reply]
@dav0000000 Setting pre-tension based on speed is inaccurate with aggressive use of the throttle. 

consider this scenario with bike setup: 
mtb, bbshd, 36T front chainring, 36T rear cog, speed 5kmph downhill slight acceleration, throttle at around 5%.

in certain scenario above engine will slam pinions repeatedly over and over, even more than once per second. there is no detection of pre-tension. somewhat hard on the pinions. engine has a hard time to catch up with the accelerating rear wheel. due to bike downhill acceleration relative gap between freewheel and engine is progressively increasing. if the maximum tension gap with a bike in stand still is around quarter of a revolution the tension gap in this bike moving scenario can be over half a revolution of the engine. speed still under 10kmph. If Im then about to start a rapid uphill in the next few meters and press the throttle to 50% it will brutally slam the pinions with hundreds of wats of power to go uphill and my speed is still at around 10-15kmph.

as long as im going uphill the rear wheel will be decelerating and the drivetrain remains in tension. then i reach top of the hill, let go of throttle and when I next engage on the flat with a speed of 7kmph it again slams the pinions.

Speed is somewhat constant in the 5 to 15kmph range the entire time while the power output varies greatly.

**Reply by kklmm** [reply]
Being able to measure engine rpm allows for a precise calculation of pre-tension time in realtime in the millisecond range by measuring supplied power, current power draw vs engine rpm deviations. Produces a smooth power output with all bike setups in all scenarios. Similar to how torque sensing engines calculate pressure on pedals in realtime in milliseconds to then immediately adjust engine power. No fixed constants. all measured in realtime.

average revolution to engage the drivetrain is quarter of the revolution of the engine on 36T chainring, 36T rear cog. in acceleration then deceleration its up to half a revolution. so delayed power output on the throttle for pre-tension for average quarter/max half a revolution should be near seamless on the throttle in under 1second.

**Reply by dav0000000** [reply]
Is it accurate to say that pretensioning is just ensuring the pawl clutch is engaged. The pawl clutch requires a firm almost perfect yet different power engagement based on current position of the pawl clutch, speed and the gear you are in. Two of those variables cannot be determined. If it's too soft then it misses the first engagement, not quite enough power and the drivetrain recoils as it engages then bounces back or too hard and it engages with a clunk. If the position of the pawl clutch is perfect so it activates immediately then it engages perfectly and you think, wow that was smooth. I find that the most important thing is to keep my chain well oiled first and foremost. For me the 1% power of pretension did not work well at speed and made funny noises. I had to increase pretension dynamically to match the speed. I pretty much use 4 percent every 15km/h so at 30km/h the current for pretension is 8% but you have to have e-brakes to turn this off and I refuse to retention automatically after releasing the e-brakes so I was left with an un-tensioned drive every time I brake. Ideally this could be further adjusted based on the motor speed but it seems like a lot of work for minimal gain. I use the same 4% every 15km/h to set a "start current" for ramp up so when ramping up from 0% current at 30km/h the first current request will be for 8% rather than 1%. You really need to hit that pawl clutch nice and smoothly with a bit of power at speed. I think the biggest improvement to the clunking drivetrain when there is no pretension is setting a start current based on speed.
